### PR TITLE
[improvement](planner) support delete from partitioned table without partition specified

### PR DESCRIPTION
**Pull request by AshinGau**
# Proposed changes

Support delete from partitioned table without partition specified in [DELETE](https://doris.apache.org/docs/sql-manual/sql-reference/Data-Manipulation-Statements/Manipulation/DELETE/) stmt.

## Usage
If it is a partitioned table, you can specify a partition. If not specified, Doris will infer partition from the given conditions. In two cases, Doris cannot infer the partition from conditions: 1) the conditions do not contain partition columns; 2) The operator of the partition column is `not in`. When a partition table does not specify the partition, or the partition cannot be inferred from the conditions, the session variable `delete_without_partition` needs to be `true` to make delete statement be applied to all partitions.

## Test case
Test case is added in `regression-test/suites/delete_p0/test_delete_from_partition.groovy`, user can delete from partitioned table without partition specified now.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 41.17 seconds
 load time: 571 seconds
 storage size: 17154602793 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221021060410_clickbench_pr_32258.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 40.23 seconds
 load time: 565 seconds
 storage size: 17154577937 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221021193408_clickbench_pr_32517.html

File: fe/fe-core/src/main/java/org/apache/doris/load/DeleteHandler.java [hunk 1]
```diff
@@ -162,14 +173,46 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
                 if (noPartitionSpecified) {
+                    // Try to get selected partitions if no partition specified in delete statement
+                    // Use PartitionPruner to generate the select partitions
                     if (olapTable.getPartitionInfo().getType() == PartitionType.RANGE
                             || olapTable.getPartitionInfo().getType() == PartitionType.LIST) {
-                        if (!ConnectContext.get().getSessionVariable().isDeleteWithoutPartition()) {
-                            throw new DdlException("This is a range or list partitioned table."
-                                    + " You should specify partition in delete stmt,"
-                                    + " or set delete_without_partition to true");
+                        Set<String> partitionColumnNameSet = Utils.execWithReturnVal(
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/DeleteHandler.java:180]: Why not just call `olapTable.getPartitionColumnNames()`?
I think we should not eliminate those exceptions.

File: fe/fe-core/src/main/java/org/apache/doris/load/DeleteHandler.java [hunk 2]
```diff
@@ -369,6 +412,77 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
+    }
+
+    // Return null if the condition is not related to the partition column,
+    // or the operator is not supported.
+    private List<Range<ColumnBound>> createColumnRange(String colName, Predicate condition) {
+        List<Range<ColumnBound>> result = Lists.newLinkedList();
+        if (condition instanceof BinaryPredicate) {
+            BinaryPredicate binaryPredicate = (BinaryPredicate) condition;
+            String columnName = ((SlotRef) binaryPredicate.getChild(0)).getColumnName();
+            if (!colName.equals(columnName)) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/DeleteHandler.java:440]:
```suggestion
            if (!colName.equalsIgnoreCase(columnName)) {
```

File: fe/fe-core/src/main/java/org/apache/doris/load/DeleteHandler.java [hunk 3]
```diff
@@ -162,14 +173,46 @@ public void process(DeleteStmt stmt) throws DdlException, QueryStateException {
...
+                        Collection<Long> selectedPartitionId = null;
+                        if (!columnNameToRange.isEmpty()) {
+                            PartitionInfo partitionInfo = olapTable.getPartitionInfo();
+                            Map<Long, PartitionItem> keyItemMap = partitionInfo.getIdToItem(false);
+                            PartitionPruner pruner = olapTable.getPartitionInfo().getType() == PartitionType.RANGE
+                                    ? new RangePartitionPrunerV2(keyItemMap, partitionInfo.getPartitionColumns(),
+                                    columnNameToRange)
+                                    : new ListPartitionPrunerV2(keyItemMap, partitionInfo.getPartitionColumns(),
+                                            columnNameToRange);
+                            selectedPartitionId = Utils.execWithReturnVal(pruner::prune);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/DeleteHandler.java:200]: Call `prune` directly?

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.47 seconds
 load time: 568 seconds
 storage size: 17154644735 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027110926_clickbench_pr_34515.html

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**